File: 91-解码方法.py
class Solution:
    def numDecodings(self, s: str) -> int:
        num1 = 1
        num2 = 1 if s[-1] != '0' else 0
        
        for i in range(len(s) - 2, -1, -1):
            if s[i] == '0':
                num = 0
            else:
                num = (num1 if int(s[i:i+2]) < 27 else 0) + num2
            num1, num2 = num2, num
        
        return num if len(s) > 1 else num2

File: 91-解码方法.cpp
class Solution {
public:
    int numDecodings(string s) {
        int n = s.size(), num1 = 1, num2, num;
        num2 = s[n - 1] == '0' ? 0 : 1;
        
        for(int i = n - 2; i > -1; i--){
            if(s[i] == '0')
                num = 0;
            else
                num = (((int)s[i] - '0') * 10 + ((int)s[i + 1] - '0') < 27) ? num2 + num1 : num2;
            num1 = num2;
            num2 = num;
        }

        return n > 1 ? num : num2;
    }
};
